File: console/src/lib/types/protocol/metrics/sample.ts
import type { CpuMetric, GpuMetric, MemoryMetric, NetworkMetric } from './system';
import type { DriveTemperatureMetric, StorageMetric } from './storage';

export interface MetricSample {
  sampleId: string;
  capturedAtMs: number;
  cpu: CpuMetric;
  memory: MemoryMetric;
  storage: StorageMetric[];
  driveTemperatures?: DriveTemperatureMetric[];
  networks: NetworkMetric[];
  gpus: GpuMetric[];
  powerWatts: number | null;
  uptimeSeconds: number;
}

File: shared/src/protocol/metrics/sample.rs
use serde::{Deserialize, Serialize};

use super::{
    CpuMetric, DriveTemperatureMetric, GpuMetric, MemoryMetric, NetworkMetric, StorageMetric,
};

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct MetricSample {
    pub sample_id: String,
    pub captured_at_ms: i64,
    pub cpu: CpuMetric,
    pub memory: MemoryMetric,
    pub storage: Vec<StorageMetric>,
    #[serde(default)]
    pub drive_temperatures: Vec<DriveTemperatureMetric>,
    pub networks: Vec<NetworkMetric>,
    pub gpus: Vec<GpuMetric>,
    pub power_watts: Option<f64>,
    pub uptime_seconds: u64,
}
